### PR TITLE
Sort shuffled indices before indexing NumpyFeed

### PR DESCRIPTION
**Pull request by swook**
It is possible to use `h5py` datasets with `tf.estimator.inputs.numpy_input_fn` as there is a high level of compatibility.

However, with `shuffle=True`, one can no longer do so due to the following error:

    TypeError: Indexing elements must be in increasing order

This small change allows for one to use large HDF5 datasets via `h5py` with shuffled batch indices.

### CODE REVIEW

**Comment by swook** [comment]
I signed it!

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
That seems to make a lot of tests fail. Could you take a look?

**Comment by swook** [comment]
@drpngx I believe this now allows tests to pass.

However, I'm unsure if the tests should be adjusted instead.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Now passes the tests.

@xiejw would have cycles to take a look?

**Comment by frankchn** [comment]
ping @xiejw

File: tensorflow/python/estimator/inputs/queues/feeding_functions.py
Comment (xiejw) [inline tensorflow/python/estimator/inputs/queues/feeding_functions.py]: The code looks good. But can you add a comment to briefly explain why this is required?

**Comment by xiejw** [comment]
Sorry for the delay. Just a nit.

**Comment by swook** [comment]
@xiejw thanks for reviewing this PR. I've added a line to explain why sorting is necessary.

**Review by xiejw — APPROVED** [review]

**Comment by xiejw** [comment]
Thanks for the contribution!

**Comment by jhseu** [comment]
The timeseries failures look related. Allen, thoughts?

**Comment by allenlavoie** [comment]
Just so I'm clear, this change should ideally only be affecting the order in which the Numpy arrays are indexed, but should be returning the same values as before down the line in TensorFlow land?

I can dig into them to figure out how exactly, but I think those test failures are indicating that with shuffle=False this PR is changing the output of the op (i.e. it's no longer returned in the input order). Probably indicates a bug?

File: tensorflow/python/estimator/inputs/queues/feeding_functions.py
Comment (xiejw) [inline tensorflow/python/estimator/inputs/queues/feeding_functions.py]: After digging into details, I think this line has a bug. You meant to put the elements into position reorder_indexes, but instead, here it is getting the elements from position reorder_indexes. I should catch this earlier. Sorry.

(I was also worrying about performance if this line does the expected assigning). 

One mini example could be:

>>> a = np.array([10, 20, 30, 40])
>>> i = [0, 1, 2, 3, 0, 1, 2, 3]                                                                                                                                                                                                                                                                                              
>>> a[i]
array([10, 20, 30, 40, 10, 20, 30, 40])
>>> sorted_integer_indexes, reorder_indexes = zip(*sorted([(value, index) for index, value in enumerate(i)]))
>>> sorted_integer_indexes = list(sorted_integer_indexes)
>>> reorder_indexes = list(reorder_indexes)
>>> reorder_indexes
[0, 4, 1, 5, 2, 6, 3, 7]
>>> a[sorted_integer_indexes][reorder_indexes]
array([10, 30, 10, 30, 20, 40, 20, 40])

**Review by xiejw — CHANGES_REQUESTED** [review]
I should catch this bug earlier. Sorry for that.

**Comment by swook** [comment]
@xiejw thanks for catching the bug out. I have committed a fix to make the tests pass. I have confirmed that tests pass on my end by rebasing on upstream.

Your concern on performance is something I agree with however...

If I understand right.. this issue occurs when wrapping around datasets to return to index 0. Perhaps it's better to split the indexing operations to allow for this.

Or dismiss this PR as it could be argued that it's a bug in `h5py`.

I leave the decision to you, the maintainers.

**Comment by xiejw** [comment]
For now, I think you can convert the h5py as numpy which is one time cost. If this has more cost than expected or it is extremely inconvenient, we can revisit this PR and propose some new ideas. 

With that said, I will close this PR and please reopen if anything changed in future. 

Thanks for your contribution!

**Comment by swook** [comment]
I understand. Not to re-open the issue, but I'd just like to mention that I'm using `h5py` for large input data which cannot be held in memory. While one could write their own pre-processing routine using tensorflow queues, using the numpy feeding functions with hdf5 structures is very convenient.

I attach the explanation for future reference.

**Pull request closed without merging**
